packages/react/src/components/form/ui-state-machine.ts

```diff
@@ -0,0 +1,343 @@
+    const onSetPassword = () => {
+      this.send({ type: "sid_ui.passwordSetReady" });
+    };
+
+    const onVerifyPassword = () => {
+      this.send({ type: "sid_ui.passwordVerifyReady" });
+    };
+
+    super(opts, {
+      status: "initial",
+      entry: () => {
+        this.sid.subscribe("passwordSetReady", onSetPassword);
+        this.sid.subscribe("passwordVerifyReady", onVerifyPassword);
+
+        this.logInFn(this.context.config, this.context.options)
+          .then((user) => {
+            if (user) {
+              this.send({ type: "sid_login.success", user });
+            } else {
+              this.send({
+                type: "sid_login.error",
+                error: new Error("User not returned from /id"),
+              });
+            }
+          })
+          .catch((error) => {
+            this.send({ type: "sid_login.error", error });
+          });
+      },
+      exit: () => {
+        this.sid.unsubscribe("passwordSetReady", onSetPassword);
+        this.sid.unsubscribe("passwordVerifyReady", onVerifyPassword);
+      },
+    });
+  }
+  send(e: Event) {
+    switch ((e as UIEvent).type) {
+      case "sid_ui.passwordSetReady":
+        this.transition(
+          createInputState("setPassword", (password) => {
+            this.sid.publish("passwordSubmitted", password);
+            this.send({ type: "sid_ui.passwordSubmitted" });
+          })
+        );
+        break;
+
+      case "sid_ui.passwordVerifyReady":
+      case "sid_ui.passwordRecovered":
+        this.transition(
+          createVerifyPasswordState(
+            (password) => {
+              this.sid.publish("passwordSubmitted", password);
+              this.send({ type: "sid_ui.passwordSubmitted" });
+            },
+            () => {
+              this.send({ type: "sid_ui.passwordRecoveryRequested" });
+            }
+          )
+        );
+
+        break;
+
+      case "sid_ui.passwordRecoveryRequested":
+        this.transition({
+          status: "recoverPassword",
+          entry: async () => {
+            try {
+              await this.recover();
+              this.send({ type: "sid_ui.passwordRecovered" });
+            } catch (e) {
+              // ignored
+            }
+          },
+        });
+        break;
+
+      case "sid_ui.passwordSubmitted":
+        this.transition({
+          status: "submitting",
+        });
+        break;
+
+      default:
+        super.send(e);
+    }
+  }
+}
+
+class DefaultUIStateMachine extends UIStateMachine<DefaultUIStatus> {
+  constructor(opts: UIStateMachineOpts) {
+    super(opts, {
+      status: "initial",
+      entry: () => {
+        this.logInFn(this.context.config, this.context.options)
+          .then((user) => {
+            if (user) {
+              this.send({ type: "sid_login.success", user });
+            } else {
+              this.send({
+                type: "sid_login.error",
+                error: new Error("User not returned from /id"),
```

Comment:
Since we're at it, it might make sense to introduce the `SlashIDError` here instead of the basic `Error` type?

Reply:
SlashIDError is a thing in the ng-evangelion, we don't have such a thing in this repo

Reply:
It is exposed through the `Errors` namespace:

```
import { Errors } from '@slashid/slashid'

new Errors.SlashIDError(// ...
```